### DISCUSSION TITLE
Support custom environment variable usage in Pipfile source URLs

### DISCUSSION THREAD

**Discussion by Badbond** [renovatebot/renovate#26447] [Suggest an Idea]
### Tell us more.

At Picnic, we have Renovate and our artifactory Nexus running self-hosted. For Pipenv, we configure our single source as follows:

```
[[source]]
name = "picnic"
url = "https://${NEXUS_USERNAME}:${NEXUS_PASSWORD}@<our-nexus-host>/repository/pypi-public/simple"
verify_ssl = true
```

All our developers' machines then have these environment variables set in their shell. To get Renovate's `pipenv` manager to work, we have to configure the `hostRule` (for lookup) and the  `customEnvVariable` (to pass them to the `containerbase` process for locking) like so:

```js
  customEnvVariables: {
    NEXUS_USERNAME: "<username-here>",
    NEXUS_PASSWORD: "<password-here>",
  },
  hostRules: [
    {
      hostType: "pypi",
      matchHost:
        "https://<our-nexus-host>/repository/pypi-public/simple/",
      username: "<username-here>",
      password: "<password-here>",
    },
  ]
```

To my understanding, with the introduction of https://github.com/renovatebot/renovate/pull/24581, the `customEnvVariables` will not be necessary anymore iff we would be using the `USERNAME` and `PASSWORD` variables instead, which get matched against the username and password of a matching `hostRule`. Unfortunately changing these variables is quite an intrusive change to update all package files and update developers' machines, let alone that one might already be using these environment variables for something else.

Given we can match `hostRule.username` and `hostRule.password` more generically based on the URL spec of basic-auth (`https://username:password@host/path`), would it be an idea to:
- Have [`findPipfileSourceUrlWithCredentials`](https://github.com/renovatebot/renovate/blob/53e1ca654c01340917c0ab1abf1991d6ad461b98/lib/modules/manager/pipenv/artifacts.ts#L89) match on any URL with basic-auth using environment variables (e.g. `https://${FOO}:${BAR}@host/path`).
- Set the `extraEnv` specific to the variables used such that `extraEnv.FOO = hostRule.username` and `extraEnv.BAR = hostRule.password`.

This would consolidate the registry configuration towards `hostRules` more.

**Comment by secustor** [comment]
IMO this is a good improvement for usability, tough it has to be ensured that only hostRules with `matchHost` are used. 
If this is no the case Renovate could leak credentials to external registries.
